dart-impl/base/src/logging.c

```diff
@@ -22,3 +39,84 @@ const int dart__base__unit_term_colors[DART_LOG_TCOL_NUM_CODES-1] = {
   DART_LOG_TCOL_RED,
   DART_LOG_TCOL_BLUE
 };
+
+static const char *loglevel_names[DART_LOGLEVEL_NUM_LEVEL] = {
+    "ERROR",
+    "WARN",
+    "INFO",
+    "DEBUG",
+    "TRACE"
+};
+
+static
+enum dart__base__logging_loglevel
+env_loglevel()
+{
+  static enum dart__base__logging_loglevel level = DART_LOGLEVEL_TRACE;
+  static int log_level_parsed = 0;
+  if (!log_level_parsed) {
+    const char *envstr = getenv(DART_LOGLEVEL_ENVSTR);
+    if (envstr) {
+      if (strcmp(envstr, "ERROR") == 0) {
+        level = DART_LOGLEVEL_ERROR;
+      } else if (strcmp(envstr, "WARN") == 0) {
+        level = DART_LOGLEVEL_WARN;
+      } else if (strcmp(envstr, "INFO") == 0) {
+        level = DART_LOGLEVEL_INFO;
+      } else if (strcmp(envstr, "DEBUG") == 0) {
+        level = DART_LOGLEVEL_DEBUG;
+      } else if (strcmp(envstr, "TRACE") == 0) {
+        level = DART_LOGLEVEL_TRACE;
+      }
+    }
+  }
+
+  return level;
+}
+
+/* GNU variant of basename.3 */
+static inline
+const char * dart_base_logging_basename(const char *path) {
+    char *base = strrchr(path, '/');
+    return base ? base+1 : path;
+}
+
+void
+dart__logging__message(
+  const char *filename,
+  int         line,
+  int         level,
+  const char *format,
+  ...
+)
+{
+  if (level > env_loglevel() ||
+      level > DART_LOGLEVEL_TRACE) {
+    return;
+  }
+  va_list argp;
+  va_start(argp, format);
+  const int maxlen = DASH__DART_LOGGING__MAX_MESSAGE_LENGTH;
+  int       sn_ret;
+  char      msg_buf[maxlen];
+  pid_t     pid = getpid();
+  sn_ret = vsnprintf(msg_buf, maxlen, format, argp);
+//  if (sn_ret < 0 || sn_ret >= maxlen) {
+//    break;
+//  }
+  dart_global_unit_t unit_id;
+  dart_myid(&unit_id);
+  // avoid inter-thread log interference
+  dart_mutex_lock(&logmutex);
```

Comment:
No it doesn't. Both `logging.h` and `mutex.h` are part of the `dart-base` interface. `logging.c` is compiled as part of `dart-base`. The modules that use the `DART_LOG*` macros will never even see the `dart_mutex_*` calls in `dart__logging__message`. So if there is no support for for threads, `logging.c` will be compiled with an empty version to `dart_mutex_*` and anything linking against `dart_base` will call that compiled function. Of course, if you decide to build `dart_base` and `dart_shmem` with different threading settings other things are likely to break...